### DISCUSSION TITLE
support exporting images displayed in Excel?

### DISCUSSION THREAD

**Discussion by fadeli1** [eclipse-birt/birt#1075] [Q&A] [answered]
birt: 4.9.0

There is a dynamic picture, which is assigned by fields in the dataset
Field Type: blob

Export PDF display images.
No pictures are displayed when you export Excel.
Does Excel support image export?

**Comment by speckyspooky** [comment]
Hello fadeli1,

yes it is an standard feature of BIRT and exists already before BIRT 4.9, therefore it could nice to know you explicit case.
I created a very small report example based on the care model with dynamic image and exported it to PDF, XLSX, Excel (XLS) - see my screen.

Please be aware the handling of images in excel is tricky because the size of the image can be set in differnt ways 
with the own hight-width settings or one of the tricks is the rounding with an own framing grid (size 1col x 1row).

Best regards

![Excel-Export-BIRT4 9](https://user-images.githubusercontent.com/41593722/193413056-30a65073-37bd-40d1-9607-e811d4371cea.png)
[cars_test.zip](https://github.com/eclipse/birt/files/9690487/cars_test.zip)

**Comment by fadeli1** [comment]
I changed from dynamic images to embedded images, exported pdf to display images, exported excel to not display images

![1](https://user-images.githubusercontent.com/1914030/194236554-c15963f1-5b33-470f-acde-e60db5b6481f.png)
![2](https://user-images.githubusercontent.com/1914030/194236582-1eb70369-7f19-434c-9e14-66e7a5aee108.png)
![3](https://user-images.githubusercontent.com/1914030/194236605-e91d1f50-65ae-454c-8fe0-3ebd591884f9.png)
![4](https://user-images.githubusercontent.com/1914030/194236622-8f6116fb-99e4-4dd4-9442-1f5ffb43165b.png)

Could it be that the Excel dependency JAR package was not imported?

**Reply by speckyspooky** [reply]
Can you get us more information about your BIRT version and do you ue the eclipse-all-in-one or rcp-designer-version?
BIRT use the excel-spudsoft-emitter to create Excel & XLSX.

My suggestion for you is to download one of the new BIRT versions 4.9 or 4.10 and try my test-report to test the excel result.

Currently I use different BIRT Versions which works perfect with Excel & images:
- BIRT 4.6 & 4.8, Java 8 - I added 2 screens of my excel-spudsoft-plugins
- BIRT 4.9 & 4.10, Java 11

![BIRT-spudsoft-emitter-0](https://user-images.githubusercontent.com/41593722/194395953-25d0fef4-3db8-4313-8807-a745ff6b3390.png)

![BIRT-spudsoft-emitter](https://user-images.githubusercontent.com/41593722/194395835-011ead4d-f6f8-473f-91a4-714ebfb95d51.png)

**Comment by fadeli1** [comment]
I use https://download.eclipse.org/birt/downloads/drops/R-R1-4.9.0-202203161719/birt-runtime-4.9.0-20220315.zip

**Reply by speckyspooky** [reply]
The BIRT version looks good.

We can try the following:
1. Please use my test report from answer above which is located below the screen - Test it with the excel output on your system
2. Can you send me please your test report like zip-file as answer and I will check your test example

The idea is to test where the problem is located your environment or thee report.

**Reply by fadeli1** [reply]
I tested the report file you provided, showing pictures in pdf or wrod, not showing pictures in xls

I continued to test whether to just put in the embedded image or export the xls without displaying the image

My runtime contains all jar packages under birt-runtime-4.9.0-20220315\WebViewerExample\WEB-INF\lib
[samplingTaskTicket.zip](https://github.com/eclipse/birt/files/9738443/samplingTaskTicket.zip)

**Reply by fadeli1** [reply]
Or you can tell me what line in the source code is going to write the image code, and I'm going to debug the source code, and it's weird that I export xls, and the cells are blank

**Reply by fadeli1** [reply]
Thank you very much for taking time out of your busy schedule to answer my questions, thank you

**Reply by fadeli1** [reply]
Do you need to set special properties to export xls? Here is some of the code I exported
```
        // set output options
        RenderOption options = new RenderOption();
        if ("text/html".equalsIgnoreCase(contentType)) { // HTML
            ...
        } else if ("application/postscript".equalsIgnoreCase(contentType)) { // Post Script
             ...
        } else if ("application/pdf".equalsIgnoreCase(contentType)) { // PDF
             ...
        } else if ("application/vnd.ms-word".equalsIgnoreCase(contentType)) { // MS Word
            ...
        }  else if ("application/vnd.ms-excel".equalsIgnoreCase(contentType)) { // MS Excel
            options.setOutputFormat("xls");
            new EXCELRenderOption(options);
        } else if ("application/vnd.ms-powerpoint".equalsIgnoreCase(contentType)) { // MS Power Point
            ...
        } else if ("application/vnd.oasis.opendocument.text".equalsIgnoreCase(contentType)) { // Open Document Text
            ...
        } else if ("application/vnd.oasis.opendocument.spreadsheet".equalsIgnoreCase(contentType)) { // Open Document Spreadsheet
             ...
        } else if ("application/vnd.oasis.opendocument.presentation".equalsIgnoreCase(contentType)) { // Open Document Presentation
             ...
        } else if ("application/vnd.openxmlformats-officedocument.wordprocessingml.document".equalsIgnoreCase(contentType)) { // MS Word 2007
             ...
        } else if ("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet".equalsIgnoreCase(contentType)) { // MS Excel 2007
            options.setOutputFormat("xlsx");
        } else if ("application/vnd.openxmlformats-officedocument.presentationml.presentation".equalsIgnoreCase(contentType)) { // MS Word 2007
             ...
        } else {
            throw new GeneralException("Unknown content type : " + contentType);
        }

        options.setOutputStream(output);
        task.setRenderOption(options);

        // run report
        Debug.logInfo("BIRT's locale is: " + task.getLocale(), module);
        Debug.logInfo("Run report's task", module);
        task.run();
        task.close();
```

**Reply by speckyspooky** [reply]
So I check further details. At frist I checked your report with BIRT 4.9 (eclipse all-in-one) and I can greate the Excel with out image.

![excel-with-image](https://user-images.githubusercontent.com/41593722/194700271-b850c059-feaf-40e2-adb7-9e134bd93408.png)

With this result we can say, that your report with embedded image is running successfully.


You use the BIRT 4.9 runtime in an integrated/embedded scenario in Java, in compare I tested all cases with the eclipse-UI.
So there is a greate difference because it seems to me that you set the full environment in you Java-integration-code.

I know in the past there was an old excel-integration which could'n handle images but the spudsoft emitter can do this.
New otpions fro testing, cann you try to set the excel-emitter explicit:

**ExcelRenderOption option=new ExcelRenderOption();
option.setOption(IRenderOption.EMITTER_ID,"uk.co.spudsoft.birt.emitters.excel.XlsxEmitter");**

Additional on it I checked my URL-calls on browser level for compare, there you can see that the XLS-format will be call with the key **"xls_spudsoft"**. Perhaps this can be also a try for your error tracking.

![excel-url-call](https://user-images.githubusercontent.com/41593722/194700556-ed02e4b5-ad8b-4860-a397-d1bf5675e924.png)

**Reply by fadeli1** [reply]
Thank you very much. Export the xls display image in the following way:
```
EXCELRenderOption option=new EXCELRenderOption(options);
option.setOption(IRenderOption.EMITTER_ID,"uk.co.spudsoft.birt.emitters.excel.XlsEmitter");
option.setOutputFormat("xls_spudsoft");
```
but there are some problems:
1. simple-master-page->page-header and page-footer are not displayed
2. multiple tables are not aligned (Perhaps you need to set all tables to a fixed width when exporting the xls)
3. page break is processed into multiple sheets (maybe this is the correct way to handle, the old XLS way is to put in a Sheet display)
4. Multiple sheets with different table widths
![55](https://user-images.githubusercontent.com/1914030/194714134-0b5ca85e-4c63-44e5-847f-c9f001eb0289.png)
![11](https://user-images.githubusercontent.com/1914030/194713544-f0855dc3-d5e2-4d47-8031-195e01ce92b6.png)
![44](https://user-images.githubusercontent.com/1914030/194713780-61bc20e7-a50f-4d3d-b34e-ea8e3d3a9ad2.png)
![22](https://user-images.githubusercontent.com/1914030/194713553-b1b4d6e5-4a59-4b73-8986-74d73719bac5.png)
![33](https://user-images.githubusercontent.com/1914030/194713670-f7cc535e-d77b-4856-9d34-64bb05412855.png)

**Reply by speckyspooky** [reply]
@fadeli1 it sounds very good that your problem is solved and you can see the image in Excel!

Before In answer your additionl question you must understand the philosophy of the excel emitter.
The philosophy isn't to create perfect styling output therefore you have many more to do like for e.g. PDF, HTML.
The main target of the developers was to get content of report into excel with focus of the report data (not of styling).

Therefor not all styling properties will be supported on Excel and you have a little bit more to do for your perfect layout.


Your questions:

- question 1. SMP-header / SMP-footer:

    The report header and report footer of the simple-page-master wount be exported to your excel-sheet.
    It will be exported to the excel-print-master page header & footer - see screen below.

- question 2.: Multiple Pages

     This is a special handling of the excle emitter, because the emitter select the table with the **higest count of columns** an aligned all other tables with this column count.
    It is a little bit like the old HTML-table handle where had 1 styling table of all.
    Your task ist that all tables must habe the same count of columns and same width. So it could be that you need some empty columns for your alignment.

- question 3.:

    You have 2 options,
    **option 1:** you can define your simple-master-page like very large "paper" with custom format
    "AND" remove the page break counter (page break register) of rows on each table
    **option 2:** You define on the report level the user property **ExcelEmitter.SingleSheet** like Boolean and set it to **true**
    So all tables will be display on 1 Sheet.
    Be aware, the table header and table row will be repeated if you use it because this option concat only the single sheets internaly.

- question 4.:

    It sound like the similar problem like question 2. If you have 1 Sheet and you create the same table layout of alle tables (due to column count topic) you will solve it.

**Image of SMP-header & SMP-footer**
![smp-header-footer](https://user-images.githubusercontent.com/41593722/194715782-7f41d5f6-6c5f-4d23-a3f3-4dee37e3cd62.png)

**Reply by hvbtup** [reply]
> My runtime contains all jar packages under birt-runtime-4.9.0-20220315\WebViewerExample\WEB-INF\lib

You are not using the web viewer obviously, so forget about the WebViewerExample directory - you can delete it!
What you need is every JAR file inside `.../ReportEngine/lib` in your class path.
Just add `.../ReportEngine/lib/*` to your class path (replace ... with the actual directory).

**Reply by fadeli1** [reply]
@speckyspooky 
hank you very much for your patience in answering my questions
I know how to deal with my problem

**Reply by fadeli1** [reply]
@hvbtup
ok, thanks

**Reply by 858882847** [reply]
我的birt导出excel也是没图像通过后台扫描birt文件导出为xls文件只有内容没有图像只有pdf格式的有图像

**Reply by 858882847** [reply]
是birt不支持excel导出图像吗？

**Reply by wimjongman** [reply]
BIRT 支持图像。请阅读此线程。

**Reply by hvbtup** [reply]
In English, please. 
Many of us aren't native speakers, so it's a bit difficult for many of us, but I think we can all read and write it well enough to understand each other. But I'm sure that most of the committers don't understand Chinese.

**Reply by 858882847** [reply]
Birt can export images through the eclipse view and as a grid table. It can export images through the back-end transmission image path to pdf files. The location of the image exported by excel is blank. How can I solve this problem? After looking at your problem, all of you said that you can export the image. Is there a lack of file or the code generated by the back-end

**Reply by 858882847** [reply]
The export format of birt is xlsx format. The width and height of the cells have changed. The contents of the cells have also shifted. It is very different from the xls format. I don't know if you have encountered this problem. The cell position of the image is blank

**Reply by 858882847** [reply]
> BIRT 支持图像。请阅读此线程。
I have updated to Birt4.13. My problem has not been solved. Now there are few online materials and I can only ask for help in the forum

**Reply by speckyspooky** [reply]
Could you please add an example of your report without external data sources but with your image-mechanism.
In best case an image of your excel result and pdf result to compare. So we are able to understand your target and implementation.
And so perhaps we can solve your issue together.

Currently what I not understand correctly try you to ad only images on excel together with data
or try you to implements excel-background-images or what is your target with the images on excel?

**Reply by 858882847** [reply]
> 您能否添加一个没有外部数据源但使用图像机制的报表示例。在最好的情况下，您的 excel 结果和 pdf 结果的图像进行比较。因此，我们能够了解您的目标和实施情况。因此，也许我们可以一起解决您的问题。
> 
> 目前我无法正确理解的是，尝试您仅将 excel 上的图像与数据一起广告，或者尝试实现 excel 背景图像，或者您对 excel 上的图像的目标是什么？

[test.zip](https://github.com/eclipse/birt/files/10987492/test.zip)

**Reply by 858882847** [reply]
> 您能否添加一个没有外部数据源但使用图像机制的报表示例。在最好的情况下，您的 excel 结果和 pdf 结果的图像进行比较。因此，我们能够了解您的目标和实施情况。因此，也许我们可以一起解决您的问题。
> 
> 目前我无法正确理解的是，尝试您仅将 excel 上的图像与数据一起广告，或者尝试实现 excel 背景图像，或者您对 excel 上的图像的目标是什么？
![T6RS)TO5I)0%FAI DL(X2~2](https://user-images.githubusercontent.com/117696273/225534647-4d18d361-8fb7-4768-85d7-743eec83ca4f.png)
The pdf file is an xls file with images. There are no images. xlsx has some images, but the format is disordered. The data is also disordered. It should be a bug

**Reply by 858882847** [reply]
[Uploading test.zip…]()

**Reply by 858882847** [reply]
> 您能否添加一个没有外部数据源但使用图像机制的报表示例。在最好的情况下，您的 excel 结果和 pdf 结果的图像进行比较。因此，我们能够了解您的目标和实施情况。因此，也许我们可以一起解决您的问题。
> 
> 目前我无法正确理解的是，尝试您仅将 excel 上的图像与数据一起广告，或者尝试实现 excel 背景图像，或者您对 excel 上的图像的目标是什么？

speckyspooky: Did you receive the file

**Comment by hvbtup** [comment]
> Could you please add an example of your report without external data sources but with your image-mechanism.

What speckyspooky means is: please attach an rptdesign file that others can test and which demonstrates the problem; not only the output.

**Comment by speckyspooky** [comment, marked as answer]
So I checked your demo-file and figured out the issue.
You use the "text"-element, configured it to HTML and used th img-tage to add the image.

This isn't the default usage of images for excel because currently the excel-Emitter doesn't support data-URL.
The data-URL will be supported currently only at the print outputs like DOCX, PDF, HTML, PPTX.

Therefore I added 2 additional cases in my demo report and the XLSX-result attached.
The effect ist for example to use embedded images on BIRT-level with the image-element
and a second BIRT-image element where I use a 2 line script (please take a look into the advanced-properties) to select your Base64-string and put it to an BIRT-image.

Please be aware I changed your scripted data set on your images to base64 & data-URL.
I hope it will help you a littel bit.

[test_20230317.zip](https://github.com/eclipse/birt/files/11006307/test_20230317.zip)

**Reply by 858882847** [reply]
Does Excel support xlsx files? Why is the difference between the xls format exported by BIRT and the xlsx format so large? Is the xlsx format chaotic? Is the xls format not chaotic? This is a bug in itself

**Reply by 858882847** [reply]
"When I debug, the xls cell is normal, and the xlsx cell is chaotic."

**Reply by speckyspooky** [reply]
This isn't chaotic. The important thing is XLS is dead because it is the very old format of excel.
The supported Microsoft format is XLSX.

The XLS file type are completely different from XLSX because XLSX is only a zip file which includes all media files in raw (like real images) and according of it the data like XML files and according style information.
Therefore you cannot compare these file types directly.

Additonal on it based on history there are 3 ways for excel:
- XLS "Excel" without spudsoft (orgininal emitter) - lowest support and functionality
- XLS "Spudsoft excel" which was sponsored from Spudsoft company to the project - standard support (but not with all features)
- XLSX "Spudsoft excel" which was sponsored from Spudsoft company to the project - standard support (but not with all features)

The both Spudsoft-excels have the same core base of implementation but the emitters are different and based on an external library wich is called apache POI.

The best way for you is to use XSLX of Spudsoft. Please be aware the excel outputs are relay good to export data but the output of formatted styled documents are very tricky because it isn't a layout optimized output it is optimized for data.

**Reply by 858882847** [reply]
![225535050-577e48dc-9d5d-4e8c-9060-c2e329350fa7](https://user-images.githubusercontent.com/117696273/226176527-24b85c97-f1b0-425b-a2cb-42d8429e4b44.png)
![225535537-87cc85c2-23a2-4c9c-837e-0cd5078331b6](https://user-images.githubusercontent.com/117696273/226176528-239538e4-8c4f-46d7-9b6c-07abc02ad5ab.png)
These two are from the same birt file. The first is xls, and the second is xlsx. Why does this happen? Is this a background configuration issue for me? Can it be resolved

**Reply by speckyspooky** [reply]
Please be aware your uestion on this discussion was, is it possible to but images to excel files.
The answer from our side is yes and I got you an example. So please start a new discussion if you have new question on excel.

The target is to avoid topic mix so that we can reference to this discussion on same topics.

Short comment to your new note, no there is no background issue.
Like my comment before we have2 options to create excel and you didn't write it.
There are many ways to optimize excel/xlsx output and so this isn't be default an issue/bug.

Please start a new discusssion if you cannot solve your issue e.g. with column-width-settings to get the same width.
And add an example without data source and chinese display names because I had to change a lot to get your test example runable.

**Reply by 858882847** [reply]
> xlsx
Have you ever tried storing images in a merge grid

**Comment by 858882847** [comment]


**Reply by wimjongman** [reply]
This discussion is closed. Please open a new one.
